File: Day19/ValidPalindrome.java
import java.util.Stack;

class ValidPalindrome {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                char lastOpen = stack.pop();
                if ((ch == ')' && lastOpen != '(') ||
                    (ch == '}' && lastOpen != '{') ||
                    (ch == ']' && lastOpen != '[')) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
    
    public static void main(String[] args) {
        ValidPalindrome solution = new ValidPalindrome();
        String s1 = "()";
        String s2 = "()[]{}";
        String s3 = "(]";
        
        System.out.println("String: \"" + s1 + "\", Is Valid: " + solution.isValid(s1));
        System.out.println("String: \"" + s2 + "\", Is Valid: " + solution.isValid(s2));
        System.out.println("String: \"" + s3 + "\", Is Valid: " + solution.isValid(s3));
    }
}

File: Day19/ValidPalindrome.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        mapping = {")": "(", "}": "{", "]": "["}

        for char in s:
            if char in mapping:
                top_element = stack.pop() if stack else '#'
                if mapping[char] != top_element:
                    return False
            else:
                stack.append(char)

        return not stack

if __name__ == '__main__':
    solution = Solution()
    s1 = "()[]{}"
    s2 = "([)]"
    print(f"String: {s1}, Is Valid: {solution.isValid(s1)}")
    print(f"String: {s2}, Is Valid: {solution.isValid(s2)}")